### PR TITLE
Various logging improvements

### PR DESCRIPTION
**Pull request by daviddrysdale**
# Checklist

- [x] Pull request affects core Oak functionality (e.g. runtime, SDK, ABI)
  - [ ] I have written tests that cover the code changes.
  - [ ] I have checked that these tests are run by [Cloudbuild](cloudbuild.yaml)
  - [ ] I have updated [documentation](docs/) accordingly.
  - [ ] I have raised an [issue](https://github.com/project-oak/oak/issues) to
        cover any TODOs and/or unfinished work.
- [ ] Pull request includes prototype/experimental work that is under
      construction.

### CODE REVIEW

**Comment by daviddrysdale** [comment]
Might be easier to review commit-by-commit.

File: oak/server/rust/oak_runtime/src/node/logger.rs [hunk 1]
```diff
@@ -60,18 +60,15 @@ impl LogNode {
...
-                        pretty_name,
-                        to_level_name(msg.level),
-                        msg.file,
-                        msg.line,
-                        msg.message
+                    Ok(msg) => log!(
+                        target: &format!("{}:{}", msg.file, msg.line),
+                        to_level(msg.level),
+                        "{}",
+                        msg.message,
```
Comment (tiziano88) [inline oak/server/rust/oak_runtime/src/node/logger.rs:71]: This makes all log lines from runtime and from the application look very similar. Do you think we should try to separate runtime and application logs at some point, perhaps redirecting output to different files?
I guess a less drastic solution would be to use more structured logging, and tag each log entry with whether it came from the runtime or the application; that way at least it may be easy to filter them out in a log viewer.
@rbehjati do you have any opinion on this? Is there a way in kubernetes to have multiple output logs?

File: oak/server/loader/oak_runner_main.cc [hunk 2]
```diff
@@ -53,6 +53,11 @@ std::shared_ptr<grpc::ServerCredentials> BuildTlsCredentials(std::string pem_roo
 int main(int argc, char* argv[]) {
   absl::ParseCommandLine(argc, argv);
 
+#ifdef OAK_DEBUG
+  // Use the failure signal handler from glog.
```
Comment (tiziano88) [inline oak/server/loader/oak_runner_main.cc:57]: Could you add a description of what this actually does? AFAICT it prints a stack trace upon receiving various signals.

File: scripts/build_server [hunk 3]
```diff
@@ -14,6 +14,7 @@ Build the specified version of the Oak Runtime server.
 
   -s    Server type:
           - base: base version of the server (default)
+          - ndebug: base version of the server with debugging compiled out
```
Comment (tiziano88) [inline scripts/build_server:17]: Should we just call these "debug" and "release", going forward? And basically mention that only the "release" binaries actually enforce IFC and are meant to be used in production.

File: oak/server/oak_runtime.cc [hunk 4]
```diff
@@ -53,7 +53,11 @@ void NodeFactory(uintptr_t data, const char* name, uint32_t name_len, uint64_t n
 std::unique_ptr<OakRuntime> OakRuntime::Create(
     const ApplicationConfiguration& config,
     std::shared_ptr<grpc::ServerCredentials> grpc_credentials) {
-  absl::call_once(glue_once, &glue_init);
+#ifdef OAK_DEBUG
+  absl::call_once(glue_once, &glue_init, 1);
```
Comment (tiziano88) [inline oak/server/oak_runtime.cc:57]: Magic number :)

File: oak/server/rust/oak_glue/oak_glue.h [hunk 5]
```diff
@@ -23,7 +23,7 @@ extern "C" {
 
 // Perform start of day initialization.  Must only be called once, before
 // any other functions.
```
Comment (tiziano88) [inline oak/server/rust/oak_glue/oak_glue.h:25]: Document that debug is interpreted as zero or nonzero and what it does.

File: oak/server/rust/oak_glue/src/lib.rs [hunk 6]
```diff
@@ -27,9 +27,11 @@ use std::{
 };
 
 #[no_mangle]
-pub extern "C" fn glue_init() {
+pub extern "C" fn glue_init(debug: u32) {
```
Comment (tiziano88) [inline oak/server/rust/oak_glue/src/lib.rs:30]: I actually don't think this should be here, we should probably have a corresponding flag / feature gate in the Rust code instead, which affects the compilation step.

File: oak/server/rust/oak_runtime/src/node/wasm/mod.rs [hunk 7]
```diff
@@ -249,13 +256,17 @@ impl WasmInterface {
...
+        );
 
         self.get_memory()
             .set_value(write_addr, write_handle as i64)
             .map_err(|err| {
                 error!(
-                    "channel_create: Unable to write writer handle into guest memory: {:?}",
-                    err
+                    "{}: channel_create(): Unable to write writer handle into guest memory: {:?}",
+                    self.pretty_name, err
```
Comment (tiziano88) [inline oak/server/rust/oak_runtime/src/node/wasm/mod.rs:269]: This seems quite annoying to maintain; could / should we have a node_{info,error,...} macro that just uses the correct prefix? We can make `Node` impl `Display` or `Debug` if it does not already, and the macro would just print `self` in that case. Can be a separate PR, if you think it's a good idea.

File: oak/server/rust/oak_runtime/src/node/logger.rs (hunk 1 above)
Comment (rbehjati) [reply oak/server/rust/oak_runtime/src/node/logger.rs:71]: GCP/kubernetes can handle logs from different resources, and the API allows writing logs to different files too. But I don't know how you could write to different files with the `log` macro. 
If that is not possible, we could tag the log entires as you suggested. I guess it is sufficient to only tag the one here.

File: scripts/build_server (hunk 3 above)
Comment (daviddrysdale) [reply scripts/build_server]: I'd avoid "debug" and "release" as labels, because they're likely to be associated with things like debug symbols, `-O ` flag values etc. (and in retrospect I should maybe have used a better name than `OAK_DEBUG`).   But yes, we should definitely make it clear that only the "logless" binaries are safe for production use with real private data.

(Moved to use `logless` as the build label for the moment.)

**Comment by daviddrysdale** [comment]
For reference, here's the difference in log output for a single `abitest` case.  Before:
```log
2020-04-17 17:37:43,974 INFO  [oak_runtime::node::logger] ThreadId(2):LogNode(logging-config) LOG: INFO examples/abitest/module_0/rust/src/lib.rs:202: running test ChannelWrite
2020-04-17 17:37:43,974 DEBUG [oak_runtime::runtime] node label: Label { secrecy_tags: [], integrity_tags: [] }
2020-04-17 17:37:43,974 DEBUG [oak_runtime::runtime] validating whether node NodeId(2) can read from channel Handle(3080333098530471254)
2020-04-17 17:37:43,974 DEBUG [oak_runtime::runtime] channel label: Label { secrecy_tags: [], integrity_tags: [] }
2020-04-17 17:37:43,974 DEBUG [oak_runtime::runtime] node NodeId(1) can write to channel Handle(8287120257203608242)
2020-04-17 17:37:43,974 DEBUG [oak_runtime::runtime] node NodeId(2) can read from channel Handle(3080333098530471254)
2020-04-17 17:37:43,974 DEBUG [oak_runtime::runtime] validating whether node NodeId(2) can read from channel Handle(3080333098530471254)
2020-04-17 17:37:43,974 DEBUG [oak_runtime::node::wasm] ThreadId(1):WasmNode(frontend-config) invoke_index: 4 RuntimeArgs([I64(9189119202715961321), I32(1453048), I32(3), I32(1), I32(0)])
2020-04-17 17:37:43,974 DEBUG [oak_runtime::runtime] node NodeId(2) can read from channel Handle(3080333098530471254)
2020-04-17 17:37:43,974 DEBUG [oak_runtime::node::wasm] ThreadId(1):WasmNode(frontend-config) channel_write: 9189119202715961321 1453048 3 1 0
2020-04-17 17:37:43,974 DEBUG [oak_runtime::runtime] wait_on_channels: channels not ready, parking thread ThreadId(2):LogNode(logging-config)
2020-04-17 17:37:43,974 DEBUG [oak_runtime::runtime] validating whether node NodeId(1) can write to channel Handle(8287120257203608242)
2020-04-17 17:37:43,974 DEBUG [oak_runtime::runtime] node label: Label { secrecy_tags: [], integrity_tags: [] }
2020-04-17 17:37:43,974 DEBUG [oak_runtime::runtime] channel label: Label { secrecy_tags: [], integrity_tags: [] }
2020-04-17 17:37:43,974 DEBUG [oak_runtime::runtime] node NodeId(1) can write to channel Handle(8287120257203608242)
2020-04-17 17:37:43,974 DEBUG [oak_runtime::node::wasm] ThreadId(1):WasmNode(frontend-config) invoke_index: 4 RuntimeArgs([I64(9189119202715961321), I32(1453064), I32(8), I32(1), I32(0)])
2020-04-17 17:37:43,974 DEBUG [oak_runtime::node::wasm] ThreadId(1):WasmNode(frontend-config) channel_write: 9189119202715961321 1453064 8 1 0
2020-04-17 17:37:43,975 DEBUG [oak_runtime::runtime] validating whether node NodeId(1) can write to channel Handle(8287120257203608242)
2020-04-17 17:37:43,975 DEBUG [oak_runtime::runtime] node label: Label { secrecy_tags: [], integrity_tags: [] }
2020-04-17 17:37:43,975 DEBUG [oak_runtime::runtime] channel label: Label { secrecy_tags: [], integrity_tags: [] }
2020-04-17 17:37:43,975 DEBUG [oak_runtime::runtime] node NodeId(1) can write to channel Handle(8287120257203608242)
2020-04-17 17:37:43,975 DEBUG [oak_runtime::node::wasm] ThreadId(1):WasmNode(frontend-config) invoke_index: 4 RuntimeArgs([I64(99999), I32(1453064), I32(8), I32(1), I32(0)])
2020-04-17 17:37:43,975 DEBUG [oak_runtime::node::wasm] ThreadId(1):WasmNode(frontend-config) channel_write: 99999 1453064 8 1 0
2020-04-17 17:37:43,975 ERROR [oak_runtime::node::wasm] channel_write: No such handle
2020-04-17 17:37:43,975 DEBUG [oak_runtime::node::wasm] ThreadId(1):WasmNode(frontend-config) invoke_index: 2 RuntimeArgs([I64(8976794370648527173)])
2020-04-17 17:37:43,975 DEBUG [oak_runtime::node::wasm] ThreadId(1):WasmNode(frontend-config) channel_close: 8976794370648527173
2020-04-17 17:37:43,975 DEBUG [oak_runtime::node::wasm] ThreadId(1):WasmNode(frontend-config) invoke_index: 2 RuntimeArgs([I64(9189119202715961321)])
2020-04-17 17:37:43,975 DEBUG [oak_runtime::node::wasm] ThreadId(1):WasmNode(frontend-config) channel_close: 9189119202715961321
2020-04-17 17:37:43,975 DEBUG [oak_runtime::runtime::channel] remove_reference: deallocating channel 22
```

After:
```log
2020-04-17 17:37:05,076 INFO  [examples/abitest/module_0/rust/src/lib.rs:203] running test ChannelWrite
2020-04-17 17:37:05,076 DEBUG [oak_runtime::node::wasm] ThreadId(1):WasmNode(frontend-config): channel_write(13594257229458153312, 1453048, 3, 1, 0)
2020-04-17 17:37:05,076 DEBUG [oak_runtime::runtime] NodeId(2): wait_on_channels: channels not ready, parking thread ThreadId(2):LogNode(logging-config)
2020-04-17 17:37:05,076 DEBUG [oak_runtime::node::wasm] ThreadId(1):WasmNode(frontend-config): channel_write(13594257229458153312, 1453064, 8, 1, 0)
2020-04-17 17:37:05,076 DEBUG [oak_runtime::node::wasm] ThreadId(1):WasmNode(frontend-config): channel_write(99999, 1453064, 8, 1, 0)
2020-04-17 17:37:05,076 ERROR [oak_runtime::node::wasm] ThreadId(1):WasmNode(frontend-config): channel_write(99999,...): No such handle
2020-04-17 17:37:05,076 DEBUG [oak_runtime::node::wasm] ThreadId(1):WasmNode(frontend-config): channel_close(16825954456857813815)
2020-04-17 17:37:05,076 DEBUG [oak_runtime::node::wasm] ThreadId(1):WasmNode(frontend-config): channel_close(13594257229458153312)
2020-04-17 17:37:05,076 DEBUG [oak_runtime::runtime::channel] remove_reference: deallocating channel 22
```
